### ISSUE TITLE
Title Bar Theming Support

### ISSUE THREAD

**Issue by RebeccaDeField** [LMMS/lmms#2450]
LMMS is built using Qt and and it uses Qt's QMdiArea (MDI stands for Multiple Document Interface). Every window inside LMMS is a QMdiSubwindow, and they are not themeable using CSS. Their look is hard-coded in the Qt settings for your PC

A solution someone mentioned would be to actually render frameless windows and then draw a custom titlebar on top of them.

I origanally added the issue here https://github.com/HDDigitizerMusic/lmms-alt-theme/issues/5, but  @Umcaruje suggested that I add this issue on the main issues tracker.

*Edited 2015-11-18 @tresf, added screenshot*
![image](https://cloud.githubusercontent.com/assets/6345473/11260986/3ef6b228-8e3e-11e5-9fac-baf35c0462fd.png)

**Comment by tresf** [comment]
Duplicate of #123, which was consolidated into meta issue #1839.

We can leave this open, but we are in dire need of C++ coders, so unless there is movement on this in the next few months, I will close this out and turn the "Modern and Sleek" thread into a meta-issue so we don't bulk up the tracker.

In regards to making our internal MDI-style windows borderless and  adding our own themed title bars, close, minimize, maximize buttons, the task is rather trivial from a seasoned Qt GUI developer, we just need some people to step up. :+1:

**Comment by BaraMGB** [comment]
I'm on it. But its not so trivial as @tresf said. But I think I can do this.


![windowdec1](https://cloud.githubusercontent.com/assets/6502580/12536764/827bd460-c2af-11e5-9ca2-6bff113a0751.png)

**Comment by Umcaruje** [comment]
@BaraMGB I love you

**Comment by tresf** [comment]
^

**Comment by RebeccaDeField** [comment]
@BaraMGB Nice work!!

**Comment by circuitfry** [comment]
Hey there, popping in to say you're the man, @BaraMGB, this is a crucial feature that themers have been needing for a long time.

**Comment by BaraMGB** [comment]
Okay, I got it. It was much more simpler as I thought at first. Since the Titel Bar belongs to the MDISubWindow widget, I only had to implement a paintEvent and draw the new Title Bar. Simple enough. The old Titel Bar is under the new and gives his functionality through the new one. So I don't need implement the hole window stuff such resize or the buttons. I only add new PNGs for close, minimize and maximize in the theme folder and load them on the right place. That's all.  

But now I have to figure out the css stuff. I don't know how to implement that. But I will see.


![windowdec2](https://cloud.githubusercontent.com/assets/6502580/12555266/9758b246-c380-11e5-88e1-4bd2cc6fc4b7.png)

**Comment by tresf** [comment]
> now I have to figure out the css stuff. I don't know how to implement that. But I will see.

Here's how I did it in the grid patch... aaa798c.  You may want to start a Pull Request now against your GitHub repo so that we can make advisement.

default/style.css

```css
qproperty-foo: rgba( red, green, blue, alpha );
```

c++

```c++

Q_PROPERTY( QBrush foo READ foo WRITE setFoo )
QBrush foo() const;
void setFoo( const QBrush & c );
QBrush m_foo
```

**Comment by RebeccaDeField** [comment]
@BaraMGB If you can give me the size of the icons for the title bars, I can design the minimize, maximize and close icons for this.

**Comment by circuitfry** [comment]
I feel the need to point out the white space suddenly appearing underneath the scrollbars. The left border seems obscured too. I am out of the loop if this is from a different problem.

**Comment by circuitfry** [comment]
Nevermind!!! Wrong screenshot!!

**Comment by tresf** [comment]
> @BaraMGB If you can give me the size of the icons for the title bars, I can design the minimize, maximize and close icons for you.

@RebeccaDeField we may go vector.  Do you agree that we may be able to tackle them all with a font + style sheet + css borders, etc?  If so, we can skip the pixmaps for these particular buttons.

**Comment by RebeccaDeField** [comment]
@tresf If we want to just code up the icons instead of using pixmaps, I could always just create a mockup to give the coders some direction for what it should look like. Let me know if that would be helpful :)

**Comment by BaraMGB** [comment]
I think @tresf meant svg instead of pixmap.

**Comment by RebeccaDeField** [comment]
@BaraMGB If so, I design everything in Inkscape so that would work as well.

Awaiting clarification from @tresf ;)

**Comment by tresf** [comment]
> I think @tresf meant svg instead of pixmap.

AFAIK we've migrated [almost no icons](https://github.com/LMMS/lmms/tree/master/data/themes/default) from PNG to SVG.  Please correct me if this assumption is wrong in regards to the new theme.

**Comment by tresf** [comment]
> @RebeccaDeField we may go vector

P.S. This statement was in regards to using pure CSS, not SVG.  Sorry for the confusion that it has caused.

**Comment by Umcaruje** [comment]
>Do you agree that we may be able to tackle them all with a font + style sheet + css borders, etc? 

I think making an icon font and/or doing css hackery for just three icons is a bit of an overkill. Also, css hackery might be fun, but using pixmaps for the close, minimize and maximize icons will mean that users can easily change them in their themes, as opposed to CSS tricks, which require a lot more knowledge.

**Comment by tresf** [comment]
> I think making an icon font and/or doing css hackery for just three icons is a bit of an overkill

Perhaps, but `_` `□` `x` in a styled box isn't rocket science either. :+1:

**Comment by RebeccaDeField** [comment]
Either way sounds great to me. Seems like @tresf could code it up pretty easily. Is there a specific reason that using css might better improve or speed up the program as opposed to just using the .pngs?

I also agree with @Umcaruje. We are low on coders so it might be a good idea to use something that can be easily updated.

Either way, I can't wait to see what it looks like in action. :)

**Comment by circuitfry** [comment]
Actually, it's funny, but I managed to read the Qt Style Sheets reference, turns out QWidget::title is how you style the title bars of all of the widgets.

**Comment by Umcaruje** [comment]
![screenshot from 2016-01-27 14 07 01](https://cloud.githubusercontent.com/assets/6282045/12614537/db4a5352-c500-11e5-922d-e1782b0c74fd.png)

I fiddled around with @BaraMGB's branch, and must say I'm pretty happy. 

Code info here: https://github.com/LMMS/lmms/pull/2516#issuecomment-175626711

**Comment by RebeccaDeField** [comment]
Great work everyone! I think we should come up with a design for the title bars that matches the Default theme and put that in place :+1:

**Comment by IvanMaldonado** [comment]
This looks sweet! I can't wait to see this in action in the 1.2. This and the other contributions you have made about theming support will open a whole new branch of themes. 

Thank you so much for doing this!

**Comment by Wallacoloo** [comment]
It occurs to me that if we were able to have LMMS' subwindows make use of the system theme, then we could save ourselves a lot of effort in the long run - the user would be able to style LMMS just by opening up their OS' appearance configurator (e.g. `Aero` for windows systems, `gnome-tweak-tool` for gnome desktops, `qtconfig` for KDE systems, etc), and we'd gain the benefit that LMMS has a consistent look-and-feel with the rest of the applications on the user's system.

A quick experiment shows for me that lmms' subwindows (the border color, minimize/close icons, background color, font color, etc) don't seem to change look at all when I alter either my system's gtk theme *or* the qt4/5 theme, regardless of if I build LMMS with `WANT_QT5` or not. I'm unsure if this is our own doing (are we explicitly overriding the system's theme in our code?) or a limitation of Qt5 and its subwindowing system. But I think it's worth looking into seeing if we can fix this on our end.

**Comment by circuitfry** [comment]
I'm not sure what you're meaning, but I *love* that we can make custom themes, and I hope we don't lose that feature. Being able to use the system's theme sounds great, *if* we can make it optional. Again, I'm not 100% on what you're saying, here, but that's how I feel.

**Comment by Wallacoloo** [comment]
@circuitfry I'll admit that I don't have a huge amount of experience programming GUIs. But as a user, I always choose applications that follow my system theme when given the choice. For example, I choose Cantata over foobar2000 as my media player because, even though the latter offers extensive internal theming capabilities, the former obeys my system's GTK theme, which makes it much less jarring when tabbing into/out of it and there's a much wider selection of stunning GTK themes out there than there are for any theming system used in only one application.

At least with GTK, the system theme provides a lot of good-looking defaults: application background color, document background color, text color, highlight color, checkboxes and sliders, *tons* of icons, window border style + window button look and arrangement. These are all complete with shadowing and optional animations (e.g. when checking/unchecking a checkbox, hovering over the window close button, etc) that are again, themeable. There's no way that we're going to be able to replicate these animations & such if we do all our theming by hand.

As an example, here's a GTK theme that I think would suit LMMS well:

![goodtheme](https://cloud.githubusercontent.com/assets/1210751/12838093/1b71c2c0-cb7e-11e5-824d-433dcf63520c.png)
It contains good-looking background colors, checkboxes, sliders, drop-down menus and highlight colors (blue and green).

But you can also do *more* than is easily possible with our current theme engine. For example, you could take Umcaruje's colored window title example & take it further by texturing it, as well as the window backgrounds, etc:

![uglyredtheme](https://cloud.githubusercontent.com/assets/1210751/12837929/460deee8-cb7c-11e5-97e0-8acd4ce80352.png)  
(Note: I dislike the above theme, but it was the first one I could find that shows color & texture)

Trying to extend Umcaruje's theme to be textured with our custom theming system would be, I imagine, more difficult & as said, animations (which the above screenshot can't illustrate) are near impossible.

There are of course lots of places where an application needs to leverage its own theme, too. For example, your system theme won't describe how to render the notes on a piano roll or how to draw a know control. *However*, it will provide a consistent color palette that you can choose from to make sure these objects look good.

My thoughts are that we should leave the basics (background colors, text colors, window borders) to the system theme when possible, and only make use of our own theming for things where the system theme doesn't provide enough context (e.g. how rounded the piano roll notes are, etc). Keep in mind that you *can* create your own system themes, so going this route actually makes theming *more* capable, due to the possibility of animations, etc. I'm not actually sure to what extent Qt can be made to follow system themes, but if it's possible, I think it should be the preferred route.

**Comment by tresf** [comment]
I'm always a fan of native look-and-feel as an option (perhaps not necessarily as a default) although  I don't believe MDI's work with native themes as the rest of the components can.

**Comment by RebeccaDeField** [comment]
We could try making the users GTK theme the default with the option of selecting a custom theme. Maybe there could be a setting somewhere with a checkbox "Use GTK theme." Chrome does something simalar to that.

Personally, I usually prefer themes that have been created manually for pograms over the ones that match. This is just my preference though, and I'm sure a lot of people would prefer it to be like what @Wallacoloo is talking about. That is why I think that having both options is important.

**Comment by circuitfry** [comment]
@Wallacoloo has a significant point, but I will also add that if we choose to rely on the system's default theme, it's going to come with some challenges. I'm not against system themes at all, I love the idea, I'm only stating for the record that it's not as simple as other applications might be. For one, LMMS uses special icons that GTK themes might not provide. We could rely on system colors, but we would need to have icon themes available to go with said GTK themes, too, ones that can be styled to match (I think SVGs actually can do this? I am no expert on SVGs).

The other challenge of course would be platform compatibility. Windows, OS X, and GNU/Linux all have their own system themes, and it's going to be a project in itself to figure out how to make all three work.

That's all that I'm outlining in the discussion. It's not a refutation or an attempt to shut down system themes. I think it can and probably should be done. These are just the obstacles that we have to get over. Artists and programmers are going to have to work together on these obstacles.

**Comment by Wallacoloo** [comment]
Yeah, it definitely seems like the consensus, and the most reasonable thing
to do here, is to continue how we are with the title bar theming. Maybe we
can target system theme options during some future UI overhall.

**Comment by BaraMGB** [comment]
Actually I need to know how to continue with this. I add some QProperys and now one can change the color of the titel bar (and text color and border color) by style.css

**Comment by circuitfry** [comment]
@BaraMGB if you ask me, I think it'd be great if we could turn the title bar into two corner background images and a middle horizontally-repeating background image connecting them together. You could simply draw these three images on top of your current CSS-able rectangle, and if themers didn't want any one of those four graphical elements, they can hide them all using display:none. Bonus points if we can make it possible to set the widths of the corners. If we did this, themers would be able to make those gimmicky techy-looking cyber-frame images that you see on Emerald themes all the time MAN I love those!

**Comment by RebeccaDeField** [comment]
@circuitfry I think that idea would compliment some of the themes greatly.

I also think that we should move forward with some simpler title bars that match the current default theme for now. Getting something clean and simple in place would greatly improve the interface.

We can definitely put something more complex and customizable in place later.

Of course, this is just my opinion on the matter :)

**Comment by IvanMaldonado** [comment]
Can I ask the status of this?
btw @Wallacoloo How is that gtk theme called? looks pretty.

**Comment by BaraMGB** [comment]
Feel free to checkout my repo and test it. 
https://github.com/LMMS/lmms/pull/2516
The color can be changed in the styles.css. This hack paints over the old title bar. Therefore you can't change the height of this. And also this corners can't be changed. I think in this case pictures are a little bit too much for this. Perhaps we can add a gradient? But please at first test this out.

**Comment by Wallacoloo** [comment]
@IvanMaldonado the first GTK Theme (with the dark colors, blue and green) is [Flat-Plat](http://gnome-look.org/content/show.php/Flat-Plat+3.14?content=167704).

**Comment by BaraMGB** [comment]
Okay, I added a gradient. The properties for the window decoration are now:

- qproperty-activeDecorationColor1
- qproperty-activeDecorationColor2
- qproperty-inactiveDecorationColor1
- qproperty-inactiveDecorationColor2
- qproperty-leftUpperBorderColor
- qproperty-titleTextColor
- qproperty-titleTextShadowColor
- qproperty-windowBorderColor
- qproperty-windowDecorationBorderColor

**Comment by Umcaruje** [comment]
@BaraMGB I think this is a wrong approach. You should make `activeDecorationColor` and `inactiveDecorationColor` properties `QBrushes` and then easily just define the gradient in CSS, like we do for [the song editor](https://github.com/LMMS/lmms/blob/master/data/themes/default/style.css#L271-L274)

Random example from the Qt Website:
```css

qlineargradient(x1: 0, y1: 0, x2: 0, y2: 1,
                                 stop: 0 #E1E1E1, stop: 0.4 #DDDDDD,
                                 stop: 0.5 #D8D8D8, stop: 1.0 #D3D3D3);
```

Another thing: why do all of these colors need to be `QProperty`'s?

The decoration color could easily be `background-color` and the text color could be `color`.
the border could also probably be done by normal css syntax. Also active windows could have their colors defined using the `:active` CSS selector, which is a valid CSS Pseudo class.

The name `titleTextColor` is also redundant to me, since there isn't any other text on the windows except the title. It should simply be `textColor`.

**Comment by Umcaruje** [comment]
Closed by #2516

**Comment by tresf** [comment]
Anyone having issues with the maximize/restore functionality of #2516?  The top left can be used to restore, the title can be double-clicked to restore, but the top right shows the maximize icon and doesn't do anything once maximized.

![screen shot 2016-05-07 at 4 13 59 pm](https://cloud.githubusercontent.com/assets/6345473/15094353/d84de78c-146e-11e6-9065-082622ce2077.png)

**Comment by BaraMGB** [comment]
@tresf I don't understand you. Can you give steps to reproduce the issue?

**Comment by tresf** [comment]
@BaraMGB disclaimer, it could be a Mac bug (I was testing a Mac-specific item when I noticed it).

1. Open a new project
2. Maximize the song-editor using the maximize button in the upper right corner of the song-editor window
3. Try using that button to restore (does't work)
 * Expected behavior, the maximize button should turn into a restore button and clicking it should restore the song-editor window.

The screenshot provided in the above post illustrates that a maximized window on Mac still shows the "Maximize" icon.  A work-around for this behavior is to double-click the title bar, or to use the context menu from the opposite corner.

**Comment by BaraMGB** [comment]
@tresf Can you provide a full screen screenshot? I think, I have a clue what's happen on mac with the title bar. On Linux the title bar disappear and these three buttons are shown in the main window menu bar. I think, on mac there isn't a main menu bar. BTW I wonder, what's happen on Ubuntu with Unity? In Unity the menu main bar looks similar to the mac. I have to check that.

**Comment by BaraMGB** [comment]
On Unity it's ugly but no bugs:
![bildschirmfoto vom 2016-05-08 13-10-47](https://cloud.githubusercontent.com/assets/6502580/15097552/12479fa4-151f-11e6-9f92-446bb6861874.png)

**Comment by tresf** [comment]
@BaraMGB what you are describing is correct, the title bar stays on Mac.  Is this something we can accommodate?

P.S. we turn off the Unity behavior in our desktop shortcut:

https://github.com/LMMS/lmms/blob/master/cmake/linux/lmms.desktop#L9

**Comment by BaraMGB** [comment]
@tresf I have tried to fix that blind. Can you test this branch? Or should I make a pull request yet?
https://github.com/BaraMGB/lmms/tree/SubWindowDecoMac

**Comment by tresf** [comment]
Quick testing.... No luck with this patch.  Doesn't seem to enter the block as the maximize button never changes to the restore button.

```diff
 		m_maximizeBtn->show();
+		if( isMaximized() )
+		{
+			m_maximizeBtn->hide();
+			m_restoreBtn->move( middleButtonPos );
+			m_restoreBtn->show();
+		}
  	}
```

**Comment by BaraMGB** [comment]
@tresf I think about it.

**Comment by BaraMGB** [comment]
Okay, next try. I implemented an own restore slot. Sorry, new branchname : https://github.com/BaraMGB/lmms/tree/SubwindowDecoMac
@tresf

**Comment by tresf** [comment]
@BaraMGB better, but still some issues... 

Using `SubwindowDecoMac` branch:
 * Click Maximize icon
   * The window properly maximizes
   * Maximize icon stays as Maximize icon
 * Click Maximize icon again (now incorrectly labelled, should be restore icon)
   * Restore icon draws at incorrect corner
 * Click the blank area where the restore icon should have drawn
   * Window restores (note, this is the second click, it should restore on the first click)

![image](https://cloud.githubusercontent.com/assets/6345473/15102181/38002270-156a-11e6-82ec-1077a8effa40.png)

**Comment by liushuyu** [comment]
>Restore icon draws at incorrect corner

I suppose it would be: https://github.com/BaraMGB/lmms/commit/fbea8a6191e12a8b1e1ffee45c007960f12e9a36#diff-ed5b7f466dc3cc99674327f65683e93cR113

It seems like the value of `move`  wasn't correct?

**Comment by BaraMGB** [comment]
@liushuyu lol yes it should be width() - 38 :-)

**Comment by tresf** [comment]
> @liushuyu lol yes it should be width() - 38 :-)

Although that fixes the position of the `restore` button, it still requires two-clicks to actually restore it.  Ideally, the restore button should appear immediately after maximized is clicked.

-Tres

**Comment by liushuyu** [comment]
>Although that fixes the position of the restore button, it still requires two-clicks to actually restore it. Ideally, the restore button should appear immediately after maximized is clicked.

First, I don't have a Mac or Unity DE, so the following is just some of my guesses.
I suppose this might be caused by a race condition... 
When the `Maximize`button is pressed, the `maximize()` function will be called: 
```CPP
void SubWindow::maximize()
{
		m_maximizeBtn->hide();
		m_restoreBtn->move( 38, 3 );
		m_restoreBtn->show();
		showMaximized();
}
```
In fact, AFAIK, two problems come from the same source: The `showMaximized()` is called *after* placing the icon, so the width and height will be incorrect: Even if we replace the fixed number 38 and 3 with variables. Then when you (@tresf in this case) press the `Maximize` button, the window maximized, but because the value of `move()` is not correct, so the button draw to the wrong place and hide. The second time that the middle button is actually the `Maximize` button again, the real `Restore` button is wrongly drawn at the corner.But since the window size is now correctly obtained (maximized status), so the middle button now becomes `Restore`.

The solution is very simple, but I don't know if it would work.
There's a small snippet:
```CPP
		showMaximized();
                m_maximizeBtn->hide();
		m_restoreBtn->move( middleButtonPos ); //Note: The variable `middleButtonPos` is not available in this scope, so you need to implement the getter yourself :-)
		m_restoreBtn->show();
```

**Comment by tresf** [comment]
> Note: The variable `middleButtonPos` is not available in this scope, so you need to implement the getter yourself :-)

Thanks for the recommendation.  I'd be happy to test once we have a working code example.

**Comment by liushuyu** [comment]
>Thanks for the recommendation. I'd be happy to test once we have a working code example.

Actually, just place this line would do the job...

```CPP
const int buttonGap = 1;
const int rightSpace = 3;
QPoint middleButtonPos( width() - rightSpace - ( 2 * m_buttonSize.width() ) - buttonGap, 3 );
```

("- Why you don't write this on the previous comment?" "- Because I am feeling lazy" 
- "Why you are feeling lazy" -"You know... ... ...")

**Comment by tresf** [comment]
Thanks.

@liushuyu to be clear, this last proposal should fix the race condition you mentioned in your previous email?

**Comment by liushuyu** [comment]
>@liushuyu to be clear, this last proposal should fix the race condition you mentioned in your previous email?

Yes, but I am unsure about whether this gonna work as expected and won't create another race condition XD

**Comment by tresf** [comment]
Thanks.  No luck, same behavior.  Pixmap is drawn as maximize instead of restore, needs to be clicked twice to restore back down.

Here's my code currently.

```diff
 		m_maximizeBtn->show();
+		if( isMaximized() )
+		{
+ 			const int buttonGap = 1;
+ 			const int rightSpace = 3;
+ 			QPoint middleButtonPos( width() - rightSpace - ( 2 * m_buttonSize.width() ) - buttonGap, 3 );
+			showMaximized();
+			m_maximizeBtn->hide();
+			m_restoreBtn->move( middleButtonPos ); //Note: The variable `middleButtonPos` is not available in this scope, so you need to implement the getter yourself :-)
+			m_restoreBtn->show();
+		}
  	}
```

**Comment by tresf** [comment]
**Edit:** Corrected code example to reflect what was tested.

**Comment by BaraMGB** [comment]
I guess, the isMaximize returns on Mac "false". That's why my very first try don't runs. And here in the resize function we ask isMaximized again for don't hiding the restore button. Therefore it hides.

**Comment by liushuyu** [comment]
Hi guys, there's what I found in Qt documents:

>Note: Due to limitations on some window systems, this does not always report the expected results (e.g., if the user on X11 maximizes the window via the window manager, Qt has no way of distinguishing this from any other resize). This is expected to improve as window manager protocols evolve.

>By default, this property is false.

No, no, Qt ,don't do this ……

**Comment by tresf** [comment]
@liushuyu I think you're confusing maximizing a Window versus maximizing a `QMdiArea`.  These are handled by Qt, not Cocoa/OS X and AFAIK they work fine without the custom code.

Here's what 1.1.3 looks like in OSX
![screen shot 2016-05-09 at 8 41 45 pm](https://cloud.githubusercontent.com/assets/6345473/15132332/7b80ec20-1626-11e6-80c4-7d4ac8605625.png)  ![screen shot 2016-05-09 at 8 41 53 pm](https://cloud.githubusercontent.com/assets/6345473/15132334/7d26a70e-1626-11e6-9a8e-5a19e0a3ec8a.png)

This default behavior works fine and I see no evidence the X11 scenario is relevant here.

**Comment by liushuyu** [comment]
>@liushuyu I think you're confusing maximizing a Window versus maximizing a QMdiArea. 

Emm, yes?

But we still need to figure out whether the `isMaximize` functions properly.

```CPP
if( !isMaximized() ) { 
qDebug() << "Maximized!";
m_restoreBtn->hide();
 }
```

If `isMaximize` works properly, you should see "Maximized!" in your terminal.

Note: you may want to `#include <QDebug>` here.

**Comment by tresf** [comment]
Shouldn't we instead listen on `changeEvent()` per:

http://doc.qt.io/qt-5/qmdisubwindow.html#changeEvent

> When the window state changes, the widget receives a `changeEvent()` of type `QEvent::WindowStateChange`.

**Comment by liushuyu** [comment]
>Shouldn't we instead listen on changeEvent() per:

Yeah! I forgot about that! Thanks for reminding me! ☺
And as usual, we need to test if Qt will properly call this callback function.

**Comment by BaraMGB** [comment]
@tresf @liushuyu I fear this event don't tell us why the event is triggered. the test is done by isMaximized() in all examples I have seen. 

I guess, we have to implement a bool and initialize that to false since all our subwindows inits not maximized. Now we can set this bool on every maximize/restore to prove the state. But that's somewhat unsave.

**Comment by liushuyu** [comment]
>@tresf @liushuyu I fear this event don't tell us why the event is triggered. the test is done by isMaximized() in all examples I have seen.

I have tried the event, and sadly this even broke all the callback functions and it totally not work.

>I guess, we have to implement a bool and initialize that to false since all our subwindows inits not maximized. Now we can set this bool on every maximize/restore to prove the state. But that's somewhat unsave.

But I came with another getter function, that is `windowState()`.

I think if that works, just do this:

```CPP
if( !( windowState().testFlag( Qt::WindowMaximized ) ) ) { m_restoreBtn->hide(); } // qDebug() <<  "triggered!" << windowState(); if any test is needed
```

**Comment by tresf** [comment]
If we can't fix this for Mac, we need to disable it altogether on that particular platform.

In the mean time, I'm happy to test changes to a fork, but I don't have the time to continuously paste the code back and fourth for this feature, I hope you understand.

**Comment by liushuyu** [comment]
>I don't have the time to continuously paste the code back and fourth for this feature, I hope you understand.

Yes, I can understand you. So I will upload my changes to another branch ASAP

**Comment by liushuyu** [comment]
>I guess, we have to implement a bool and initialize that to false since all our subwindows inits not maximized. Now we can set this bool on every maximize/restore to prove the state. But that's somewhat unsave.

@BaraMGB, we can make this an alternative plan if all the other solutions were failed.

**Comment by BaraMGB** [comment]
I'm a little bit busy in real life at the moment. I'll look at this at weekend. @tresf I'm shure we bring that thing to run for Mac, too :beers:

**Comment by BaraMGB** [comment]
@tresf : Please test this branch: https://github.com/BaraMGB/lmms/tree/SubwindowDecoMac. I have implemented the suggestion of @liushuyu ( `windowState().testFlag( Qt::WindowMaximized) `). On Maximize the window it should be there an output on console: "maximize!". To implement this whole stuff on mac could be more difficult if that fails. That could be a problem.

**Comment by tresf** [comment]
I'm not sure what I'm testing, but the maximize button does nothing after it has been clicked once.  It maximizes, never switches to restore and never works after that.  There is nothing in the console that says `maximize!`.

**Comment by liushuyu** [comment]
@tresf Then this meant that the condition ` windowState().testFlag( Qt::WindowMaximized) ` never turned True, which means Qt cannot handle the window status under Mac... That's really unfortunate

**Comment by tresf** [comment]
Is there a reason we've no `changeEvent` handlers?  Am I missing something fundamental here?  Why are we using the constructor to detect a flag?

**Comment by liushuyu** [comment]
Hmm... I've tried the `changeEvent()` , but it proved to break everything.
Somehow the window itself will misbehave when I even only add the dummy
`changeEvent()` (empty function) . And as for calling the constructor, that
was a hack which ran just fine on Linux and Windows, however, I was
disappointed when you told me this still didn't work.

**Comment by tresf** [comment]
> I've tried the `changeEvent() ` , but it proved to break everything.

In your tests, did you re-fire the super-class method?  (Does C++ do this automatically on override or do you lose it without explicit super-class invocation?)

**Comment by tresf** [comment]
@liushuyu yeah, from what I'm reading, this would be an explicit override and would require calling `QWidget::changeEvent(event);` somewhere in the function as to reimplement the now overridden behavior.

http://www.qtcentre.org/archive/index.php/t-48140.html

**Comment by liushuyu** [comment]
Oh... Thanks for reminding me that!
But the problem here is Qt couldn't handle this kind of event on Mac , as
it don't know whether the Window status changed when Maximized.

**Comment by tresf** [comment]
> But the problem here is Qt couldn't handle this kind of event on Mac , as
it don't know whether the Window status changed when Maximized.

I'm not sure I see any evidence of that in the previous code examples.  Checking on constructor is going to trigger false every time, no?

**Comment by liushuyu** [comment]
>I'm not sure I see any evidence of that in the previous code examples. Checking on constructor is going to trigger false every time, no?  

Hmm, in fact, `changeEvent` and `resizeEvent`both are working on Linux and Windows ... In addition `Qt::WindowMaximized` is a constant not a constructor. And I tested this trick before on Linux (my laptop) and Windows (my friend's) , this just worked

Anyway, thanks for your hard work, intensive reading and testing!!! :+1:

**Comment by BaraMGB** [comment]
@tresf it seems, Qt don't really maximize the subwindow on Mac. It will only resized to the max. Therefore the flags, isMaximized and Qt::WindowMaximized, return false. 

I have to write a workaround now for this. But this is much more difficult. Because I have to implement new slots, new double click event and new context menu. We have to proof all maximize order go thru our own slot to set a own flag. 

Unfortunately that doesn't runs on non Mac systems, because there we have no chance to connect the restore button with our own slot. 

I do my best to make the code as simple as possible.

**Comment by tresf** [comment]
I'm not convinced as restore works just fine from the context menu.

**Comment by BaraMGB** [comment]
Yes, but we have to know if the user restores the window for change the buttons. The context menu don't tell us. So we have to implement an own context menu.

**Comment by tresf** [comment]
> Yes, but we have to know if the user restores the window for change the buttons. The context menu don't tell us. So we have to implement an own context menu.

I feel we're over-complicating this.  To state there's no way to know that the context menu has been used I feel to be a false statement.  Furthermore, to design this around the context menu I find to be a bad idea as well.  We should detect the resize event and act accordingly.

**Comment by BaraMGB** [comment]
@tresf Okay, I've figured out an other way, perhaps. But I need your help again. If we check the size of the subwindow in relation to the MdiArea, so there is a way to check if the subwindow is maximized. I have add a qDebug() which gives us the size of both. My plea to you today is to start this one, maximize the songeditor and paste the console output here. I can write a little function to test if the subwindow is maximized on mac and all is fine. 

Here is the branch: https://github.com/BaraMGB/lmms/tree/fixMacSubwindow

Thank you for your efforts.

**Comment by tresf** [comment]
Output:

```
Area size:  QSize(655, 448) Window size:  QSize(543, 335) 
Area size:  QSize(655, 448) Window size:  QSize(400, 200) 
Area size:  QSize(655, 448) Window size:  QSize(600, 300) 
Area size:  QSize(997, 591) Window size:  QSize(983, 577) 
Area size:  QSize(997, 591) Window size:  QSize(997, 591) 
Area size:  QSize(997, 591) Window size:  QSize(600, 300) 
```

**Comment by liushuyu** [comment]
>```
Area size:  QSize(655, 448) Window size:  QSize(543, 335) 
Area size:  QSize(655, 448) Window size:  QSize(400, 200) 
Area size:  QSize(655, 448) Window size:  QSize(600, 300) 
Area size:  QSize(997, 591) Window size:  QSize(983, 577) 
Area size:  QSize(997, 591) Window size:  QSize(997, 591) 
Area size:  QSize(997, 591) Window size:  QSize(600, 300) 
```

@tresf For what I see, it seems that you only maximized the window only once (the 6th value), is that correct?

**Comment by tresf** [comment]
> @tresf For what I see, it seems that you only maximized the window only once (the 6th value), is that correct?

I maximized the song editor and then had to use the context menu to restore it.  I don't know which line is which, but at a glance, I'd assume it's the 5th where `Area size = Window size`

**Comment by liushuyu** [comment]
>I maximized the song editor and then had to use the context menu to restore it. I don't know which line is which, but at a glance, I'd assume it's the 5th where Area size = Window size

Then, I assume that means this will ... work ... ? (Finally...)

**Comment by tresf** [comment]
> Then, I assume that means this will ... work ... ? (Finally...)

Yeah.  I just tried it again.  Although I don't entirely trust the 4th dimensions much (shouldn't they be identical to the 6th?) I can confirm the 5th line is what happens when I maximize.

**Comment by BaraMGB** [comment]
I think we got a winner. The window size is identical to the MdiArea size. 👍

**Comment by BaraMGB** [comment]
@tresf Okay, I've updated the branch. I cross my fingers but this should fix the issue.

**Comment by tresf** [comment]
Success.  Tested with SongEditor and PianoRollEditor.

![image](https://cloud.githubusercontent.com/assets/6345473/15543946/daecb9d8-2264-11e6-9bcf-67ab5d091391.png)

**Comment by liushuyu** [comment]
>Success. Tested with SongEditor and PianoRollEditor.

That's amazing!! Thanks @tresf @BaraMGB for your hard work!! Congratulations!!

**Comment by BaraMGB** [comment]
Great! I'll make a pull request tonight.

**Comment by tresf** [comment]
> Great! I'll make a pull request tonight.

Merged, thanks!
